SpatialAcceleration class

amcastro-tri opened the pull request:
In preparation for MBT acceleration kinematics, this PR introduces the `SpatialAcceleration` class. This class is in the same family of spatial vectors as `SpatialVelocity` and `SpatialForce`. Of particular importance is the `SpatialAcceleration::Shift()` operation which allow us to compute the transformation of spatial accelerations of frames in a rigid body.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/6429)
<!-- Reviewable:end -->

amcastro-tri commented:
Evan, you reviewed the previous spatial algebra PR's and so it makes sense you review this one. +@edrumwri please.

---

Review status: 0 of 7 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-Kn_wYeAB15-DjdVUrGx:bdei6l9)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
This mostly looks good, Alejandro. The rationale behind many of the statements in the SpatialAcceleration class is missing, though I think if you address my comments, that rationale should start to emerge.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knd_NW4EP0PF-hoRinL:-Knd_NW4EP0PF-hoRinM:bspwf9z) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L16)):*
> ```objc
> /// rotational (angular acceleration) and translational (linear acceleration)
> /// components. Spatial accelerations are 6-element quantities that are pairs of
> /// ordinary 3-vectors. Elements 0-2 constitute the angular acceleration
> ```

Alejandro, why talk about the data layout here? Can't that discussion be delayed until considering getting/setting the spatial acceleration to/from a vector?

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kndh3-XNACWAUIcchIe:-Kndh3-YNNxkaPcs3YfO:bv3n1ih) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L23)):*
> ```objc
> /// frame X. That is @f$^XA^Y = \frac{^Xd ^XV^Y}{dt} @f$ where
> /// @f$\frac{^Xd}{dt} @f$ denotes the time derivative taken in frame X. In
> /// source code comments we write the previous expression as `A_XY = d/dt V_XY`.
> ```

I don't understand the need for the distinction here- why is it necessary to create new notation for the time derivative of `V_XY`? My ignorance is likely due to the opacity of the operations that are involved, though, if I don't understand the need for the distinction, other readers are unlikely to. Can you show the need for the distinction mathematically?

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne11WYR-GXM4l134SO:-Kne11WZccyZuD8d0nm3:bn8ir7g) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L29)):*
> ```objc
> /// the time derivative is taken is the "measured-in" frame.
> /// As with SpatialVelocity, the two contained vectors must be expressed in the
> /// same "expressed-in" frame E, which may be distinct from either X or Y.
> ```

Is this why the special distinction is necessary (i.e., that the "expressed in" frame) is different?

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1q2hhvOjDRW45MMB:-Kne1q2hhvOjDRW45MMC:b7jd95k) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L30)):*
> ```objc
> /// As with SpatialVelocity, the two contained vectors must be expressed in the
> /// same "expressed-in" frame E, which may be distinct from either X or Y.
> /// Finally, while angular acceleration is identical for any frame fixed to a
> ```

Correct only for "measured in", not expressed in, yes?

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1xRz786d13-90mys:-Kne1xRz786d13-90myt:b3t1qvl) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L31)):*
> ```objc
> /// same "expressed-in" frame E, which may be distinct from either X or Y.
> /// Finally, while angular acceleration is identical for any frame fixed to a
> /// rigid body, translational acceleration refers to a particular point. Only
> ```

I'm still left with a sense of ambiguity after reading this statement. To clarify, you mean "refers to a particular point affixed to a rigid body"? 

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 67 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbVVRcZUfuasWgg2ST:-KnbVVRcZUfuasWgg2SU:b1ok4o6) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L67)):*
> ```objc
>                       const Eigen::Ref<const Vector3<T>>& a) : Base(alpha, a) {}
> 
>   /// SpatialAcceleration constructor from an Eigen expression that represents a
> ```

This seems like the ideal place to talk about the layout of the underlying spatial acceleration, assuming that you actually need to do that. If you don't, I'd just say here that the first three components of the six dimensional vector refer to angular acceleration and the next three components refer to linear acceleration.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneC03wdTFcdmN2atjU:-KneC03wdTFcdmN2atjV:b-1smegw) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L77)):*
> ```objc
>   /// In-place shift of this spatial acceleration `A_FP` of a frame P into the
>   /// spatial acceleration of another frame Q instantaneously moving together
>   /// with frame P.
> ```

Would you express this mathematically for me? In other words, does this mean that V_FP = V_FQ?

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneFAOOnVlGOoKQXCvL:-KneFAOOnVlGOoKQXCvM:bm4p0mf) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L82)):*
> ```objc
>   /// spatial velocity V_PQ being zero at all times.
>   ///
>   /// The shift operation modifies `this` spatial acceleration `A_FP_E` of a
> ```

Would you mind showing the underlying math here too (i.e., the derivation)? *I* have a good idea of why you need to account for these changes, but only because I'm going from memory. Still, I have to take your word that your "shift" is not leaving out a quantity to be transformed, which is not good for the quality of my review.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 94 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneFgvOpNucHIbUd2dF:-KneFgvOpNucHIbUd2dG:b-vg5tsy) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L94)):*
> ```objc
>   /// of frame Q as the position vector `p_PoQo_E` expressed in the same frame E
>   /// as `this` spatial acceleration. The operation performed, in
>   /// coordinate-free form, is:
> ```

What does "coordinate-free form" mean? You mean you are omitting the expressed-in frame from the notation?

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 97 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneG_baJ0VHM1iz9NrX:-KneG_baJ0VHM1iz9NrY:bqd86f6) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L97)):*
> ```objc
>   /// <pre>
>   ///   alpha_FQ  = alpha_FP,  i.e. the angular acceleration is unchanged.
>   ///   a_FQo = a_FPo + alpha_FP x p_PoQo + w_FP x w_FP x p_PoQo
> ```

Symbols here do not match code below. Please update to make verifying your implementation easier.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 113 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneGNC3t6B8cCzKeG_a:-KneGNC3t6B8cCzKeG_b:b-vy0xu6) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L113)):*
> ```objc
>   ///   acceleration, and E must be the same expressed-in frame as for this
>   ///   spatial acceleration.
>   /// @param[in] w_FP_E
> ```

Documentation does not match your parameter name below.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneGomv3uPtIKXGFpPz:-KneGomv3uPtIKXGFpQ-:bo51cet) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L145)):*
> ```objc
>   /// does not change the original object. See ShiftInPlace() for more
>   /// information.
>   ///
> ```

Why not just omit all of the docs below and say "see ShiftInPlace() for a description of the parameters"? That should improve code maintainability.

---

*[drake/multibody/multibody_tree/math/spatial_vector.h, line 178 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbSH00u-Y0gQ22yRVE:-KnbSH00u-Y0gQ22yRVF:bd8jymp) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_vector.h#L178)):*
> ```objc
>   /// translational components are both zero.
>   static SpatialQuantity Zero() {
>     return SpatialQuantity{}.SetZero();
> ```

Ok, this is new to me- what is this `SpatialQuantity{}`?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbTgJ0uRuT8tE3mhTZ:-KnbTgJ0uRuT8tE3mhT_:besobi) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L33)):*
> ```cpp
> };
> 
> // traits specialization for SpatialForce.
> ```

BTW What is this for?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 349 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHgm6ImW895fUscwg:-KneHgm6ImW895fUscwh:b-q55kxy) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L349)):*
> ```cpp
> // think of frames P and A having coincident origins.
> // A third frame Q translated by a position p_PoQo moves rigidly with P.
> // The angular velocity w_AP_E is as a component out of plane in the z-axes
> ```

BTW What do you mean by "as a component out of plane in the z-axes"?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 357 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneH5G7KLFzbCD-zLMl:-KneH5G7KLFzbCD-zLMm:b-q894vl) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L357)):*
> ```cpp
>   typedef typename TestFixture::ScalarType T;
> 
>   // The spatial acceleration of frame P measure in A is zero.
> ```

BTW "measure" -> "measured"

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 369 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHHtDHbF-H-7jZmGr:-KneHHtDHbF-H-7jZmGs:biann0w) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L369)):*
> ```cpp
> 
>   SpatialAcceleration<T> A_AQ_expected;
>   A_AQ_expected.rotational() = Vector3<T>::Zero();
> ```

This test will only work if your derivation is correct. Is there a published problem, with known result (perhaps from a mechanics text), that you can draw upon to get an extra measure of security? 

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 378 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHvY65CJhbkguUjNa:-KneHvY65CJhbkguUjNb:be075zi) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L378)):*
> ```cpp
> }
> 
> // Unit test for the method SpatialAcceleration::Shift().
> ```

Same concern with the test above- can you use a published problem result instead?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 418 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneIAdup79rRHSrtjRi:-KneIAdup79rRHSrtjRj:b-4pf84m) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L418)):*
> ```cpp
> // This test is a combination of Case 1 (with centrifugal acceleration), Case 2
> // (with angular acceleration) and a translational acceleration a_AP of frame P
> // in A.
> ```

Same issue with this test as with those above.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-Kne1FQXFDlq-oE2OZ8p:b-3ebhuo)*
<!-- Sent from Reviewable.io -->

jamiesnape commented:
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-everything please
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-memcheck-lsan please

amcastro-tri commented:
Thank you for the review @edrumwri. Notice I merged with #6451 (under review) so you could see how these come together. Hopefully that helps to understand the notation on time derivatives which is a must for acceleration kinematics.

---

Review status: 4 of 11 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knd_NW4EP0PF-hoRinL:-KnkjNa0_K0NJp9aZOM_:b-l7twyc) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L16)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Alejandro, why talk about the data layout here? Can't that discussion be delayed until considering getting/setting the spatial acceleration to/from a vector?
</blockquote></details>

This discussion came up when implementing SpatialVector/SpatialVelocity. Remember? this documentation is the result of almost two months of discussion. Most likely these are either @sherm1's or your words. I just followed the template we decided on back then.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 23 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kndh3-XNACWAUIcchIe:-Knkjj3YqIGLEUcA4lm1:b-kksyko) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L23)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I don't understand the need for the distinction here- why is it necessary to create new notation for the time derivative of `V_XY`? My ignorance is likely due to the opacity of the operations that are involved, though, if I don't understand the need for the distinction, other readers are unlikely to. Can you show the need for the distinction mathematically?
</blockquote></details>

As you might have seen there is an on-going email discussion. This discussion lead to `ShiftTimeDerivative()` (well, not really, the discussion lead to me pushing it sooner, I actually already had it in my branch) and also to the new `SpatialAcceleration::ShiftTimeDerivative()` in this PR, see unit tests.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne11WYR-GXM4l134SO:-Knkk2nwaM_BWAZaXb1q:b-nl3qgq) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L29)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Is this why the special distinction is necessary (i.e., that the "expressed in" frame) is different?
</blockquote></details>

No, see the `ShiftTimeDerivative()` methods and unit tests first. Those are two different things. Also, take a look at `operator+()` for spatial acceleration.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 77 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneC03wdTFcdmN2atjU:-KnkmFu4wULDuCHBhDxn:bmu8m80) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L77)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Would you express this mathematically for me? In other words, does this mean that V_FP = V_FQ?
</blockquote></details>

No, those are related by `V_FQ_E = V_FP_E.Shift(p_PoQo_E)`

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 94 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneFgvOpNucHIbUd2dF:-KnknjVUonpcrImU38T6:buwumzn) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L94)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

What does "coordinate-free form" mean? You mean you are omitting the expressed-in frame from the notation?
</blockquote></details>

vectors are invariants and they are objects that do make sense even before we express them in any particular frame. So yes, that is what I mean.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 97 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneG_baJ0VHM1iz9NrX:-Knko54KtvIFCjfUEIYq:bb3xcu9) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L97)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Symbols here do not match code below. Please update to make verifying your implementation easier.
</blockquote></details>

Done, updated code.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 113 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneGNC3t6B8cCzKeG_a:-KnkoXQBQMjVQobfBoHX:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L113)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Documentation does not match your parameter name below.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 145 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneGomv3uPtIKXGFpPz:-Knkol76UFvaisvOAHCQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L145)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Why not just omit all of the docs below and say "see ShiftInPlace() for a description of the parameters"? That should improve code maintainability.
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/math/spatial_vector.h, line 178 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbSH00u-Y0gQ22yRVE:-KnkombvnoECWlvRE1wN:baprq1y) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_vector.h#L178)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Ok, this is new to me- what is this `SpatialQuantity{}`?
</blockquote></details>

`Ctrl + B` in CLion. It's a typedef. This is a CRTP class remember?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbTgJ0uRuT8tE3mhTZ:-Knkoxib1FKFAEXL2qj8:b-uko7k0) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L33)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW What is this for?
</blockquote></details>

BTW, this is just boiler plate to write the `TYPED_TEST_CASE`. Fetch the code, open it up in CLion, and you'll see all other tests (already in master) are written in the same way.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 349 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHgm6ImW895fUscwg:-KnkpJhgdH672EIcKaBH:brbgjli) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L349)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW What do you mean by "as a component out of plane in the z-axes"?
</blockquote></details>

BTW, Done

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 369 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHHtDHbF-H-7jZmGr:-KnkpQlpveoOWUqpLAnG:b-xflib1) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L369)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

This test will only work if your derivation is correct. Is there a published problem, with known result (perhaps from a mechanics text), that you can draw upon to get an extra measure of security? 
</blockquote></details>

It is basic kinematics, not really. Maybe some undergraduate book on physics? I don't think worth it. The trick here is that I do a "different" math here. I obtained these answers just by pure geometric (rather than differential) reasoning. Therefore the chances of getting the exact same result are really pretty low.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 378 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHvY65CJhbkguUjNa:-Knkpj1TvUtWvh7zM6Fk:bhwv89z) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L378)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same concern with the test above- can you use a published problem result instead?
</blockquote></details>

I cannot think of a basic text for this. Maybe the classic Resnick Halliday on an introductory course on physics?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 418 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneIAdup79rRHSrtjRi:-Knkpz89siJHvmFZg3yA:b2kedel) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L418)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Same issue with this test as with those above.
</blockquote></details>

same answers. These derivations are (mostly) purely geometrical and do not perform the same operations as the `Shift()` operations. The chances of getting a match are real low.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-Knkr7-MBmUuE1r33zaN:b7tljbv)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 4 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneFAOOnVlGOoKQXCvL:-KnkvrYvA4iPXTBqUMb6:b-2agm5p) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L82)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Would you mind showing the underlying math here too (i.e., the derivation)? *I* have a good idea of why you need to account for these changes, but only because I'm going from memory. Still, I have to take your word that your "shift" is not leaving out a quantity to be transformed, which is not good for the quality of my review.
</blockquote></details>

The derivation? I don't think that'd belong here. Also, you don't have to re-derive the math to make a review, that'd mean you are re-dong my work, which I already did. I think your review should trust I did my work here and you just make sure this is in a usable shape with no major problems?
I'll see what I can do, but keep that in mind.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 2 of 7 files at r2.
Review status: 6 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneFAOOnVlGOoKQXCvL:-KnkxPefb6h52gUc0xct:b6ykeeb) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L82)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

The derivation? I don't think that'd belong here. Also, you don't have to re-derive the math to make a review, that'd mean you are re-dong my work, which I already did. I think your review should trust I did my work here and you just make sure this is in a usable shape with no major problems?
I'll see what I can do, but keep that in mind.
</blockquote></details>

That's a bullshit response, Alejandro. If you want me to thoroughly check this out, I need something more to go by than just some hand waving.

---

*[drake/multibody/multibody_tree/math/spatial_vector.h, line 178 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbSH00u-Y0gQ22yRVE:-KnkuvoA7f24JsqJKmne:bumi76r) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_vector.h#L178)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

`Ctrl + B` in CLion. It's a typedef. This is a CRTP class remember?
</blockquote></details>

No, I mean the braces operator.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 369 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHHtDHbF-H-7jZmGr:-KnkwEPEERZooYGfSWZh:bw63w0l) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L369)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

It is basic kinematics, not really. Maybe some undergraduate book on physics? I don't think worth it. The trick here is that I do a "different" math here. I obtained these answers just by pure geometric (rather than differential) reasoning. Therefore the chances of getting the exact same result are really pretty low.
</blockquote></details>

Fair point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 4 of 11 files reviewed at latest revision, 17 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 229 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnkxyoTJEobhh6XIW50:-KnkxyoTJEobhh6XIW51:b7b1k3j) ([raw file](https://github.com/robotlocomotion/drake/blob/97ad6c6d137363e1c1f3e37d94afcf64917901c2/drake/multibody/multibody_tree/math/spatial_acceleration.h#L229)):*
> ```objc
> /// The first term is obtained by shifting the spatial acceleration `A_EA` to B
> /// as: <pre>
> ///   DtE(V_EAb) = A_EAb = A_EA.Shift(p_AB, w_EA)
> ```

I believe this is wrong, though @mitiguy will have to provide the definitive verdict. The problem is that the shift operator provides the acceleration of an offset point Ab in frame A (that is, a point fixed in A but coincident with B), but the derivative must include the fact that point B was moving around on A at the time we decided where to put point Ab.

Another way to look at this is that the original expression was 
``` 
V_EB = V_EA.Shift(p_AB) + V_AB
```
so 
```
d/dt V_EB = d/dt V_EA.Shift(p_AB) + d/dt V_AB
```
and we need the time derivative of the Shift operator, not just the time derivative of the result.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 6 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 229 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnkxyoTJEobhh6XIW50:-Knl2SB4CU-uwqNX1JR3:bhpg7we) ([raw file](https://github.com/robotlocomotion/drake/blob/97ad6c6d137363e1c1f3e37d94afcf64917901c2/drake/multibody/multibody_tree/math/spatial_acceleration.h#L229)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

I believe this is wrong, though @mitiguy will have to provide the definitive verdict. The problem is that the shift operator provides the acceleration of an offset point Ab in frame A (that is, a point fixed in A but coincident with B), but the derivative must include the fact that point B was moving around on A at the time we decided where to put point Ab.

Another way to look at this is that the original expression was 
``` 
V_EB = V_EA.Shift(p_AB) + V_AB
```
so 
```
d/dt V_EB = d/dt V_EA.Shift(p_AB) + d/dt V_AB
```
and we need the time derivative of the Shift operator, not just the time derivative of the result.
</blockquote></details>

One might then reasonably ask ... why do the unit tests pass? Speculation: the missing term above (if it is actually missing) results in the final translational acceleration containing a term ω×v when it should contain 2ω×v. If v is zero or perpendicular to ω in the unit test the missing factor of 2 wouldn't matter. Just saying ...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 82 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneFAOOnVlGOoKQXCvL:-KnlDhGI-8gntf_VkFPb:bjjbw1s) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L82)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

That's a bullshit response, Alejandro. If you want me to thoroughly check this out, I need something more to go by than just some hand waving.
</blockquote></details>

It's just a few lines in paper. I added it below for your entertainment. This plus the `ShiftTimeDerivative()` explain the `operator+()`

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 229 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnkxyoTJEobhh6XIW50:-KnlDxhnLEzMEKWNdsKa:b-nzzqy7) ([raw file](https://github.com/robotlocomotion/drake/blob/97ad6c6d137363e1c1f3e37d94afcf64917901c2/drake/multibody/multibody_tree/math/spatial_acceleration.h#L229)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

One might then reasonably ask ... why do the unit tests pass? Speculation: the missing term above (if it is actually missing) results in the final translational acceleration containing a term ω×v when it should contain 2ω×v. If v is zero or perpendicular to ω in the unit test the missing factor of 2 wouldn't matter. Just saying ...
</blockquote></details>

I added derivation in the `Shift()` operator. That derivation and `ShiftTimeDerivative()` define all the operations needed to understand `operator+()`. Thinking in terms of  Jain's shift operator is a real bad idea because it adds a lot of extra complexity. If you wonder where the derivative of the shift operator is just look at `SpatialAcceleration::Shift()` which essentially implements the time derivative of `SpatialVelocity::Shift()`. See the derivation I just pushed for `SpatialAcceleration::Shift()` .

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: 6 of 11 files reviewed at latest revision, 16 unresolved discussions, some commit checks failed.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 229 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnkxyoTJEobhh6XIW50:-KnlINFvvFghK2q6FTaR:b-i6egwv) ([raw file](https://github.com/robotlocomotion/drake/blob/97ad6c6d137363e1c1f3e37d94afcf64917901c2/drake/multibody/multibody_tree/math/spatial_acceleration.h#L229)):*
> The problem is that the shift operator provides the acceleration of an offset point Ab in frame A (that is, a point fixed in A but coincident with B), but the derivative must include the fact that point B was moving around on A

Yeap, you are right on this one. I didn't realize `SpatialAcceleration::Shift()` assumes `V_AB = 0` which defeats the purpose here. That would lead to the traditional 2ω×v term as we always see it in textbooks. 

I guess, as Paul mentioned, having `operator+()` here really is a bad idea. Maybe better to have a method `SpatialAcceleration::ShiftToMovingFrame()` that gets the `V_AB` factor right?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
All right, @sherm1, @mitiguy and @edrumwri. I added method `ComposeWithMovingFrameAcceleration()` with all derivations in there and removed the `operator+()`. PTAL!

---

Review status: 2 of 7 files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-KnoiFlcgBQ-Mc5S7cAr:btyz20e)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
Reviewed 1 of 7 files at r2, 1 of 7 files at r3.
Review status: 4 of 6 files reviewed at latest revision, 14 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knd_NW4EP0PF-hoRinL:-Knl1xFDkvAkRN5Q57N_:bc5ntxg) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L16)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

This discussion came up when implementing SpatialVector/SpatialVelocity. Remember? this documentation is the result of almost two months of discussion. Most likely these are either @sherm1's or your words. I just followed the template we decided on back then.
</blockquote></details>

I don't remember settling on any such wording for this. Can you justify it better than "that was the best we came up with six months ago"?

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbTgJ0uRuT8tE3mhTZ:-Knp0Eq-6HeuhioQxLGT:b-ge9xsv) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L33)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, this is just boiler plate to write the `TYPED_TEST_CASE`. Fetch the code, open it up in CLion, and you'll see all other tests (already in master) are written in the same way.
</blockquote></details>

BTW Would you mind giving it a better explanation? If I've not been reviewing similar code (even though I must have, I don't recall this, and you have to expect the average reader of this code will be seeing this for the first time), I have no idea what a "traits specialization" is or why it's appearing in this file.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 349 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHgm6ImW895fUscwg:-Knp0iZ3-RWEIr1VD-c4:bh800ir) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L349)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

BTW, Done
</blockquote></details>

BTW Ok, now I'm being pedantic, so would you please say instead "the angular velocity vector w_AP_E is orthogonal to the x/y plane"? 

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 418 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneIAdup79rRHSrtjRi:-Knl1o5n0u6GiFyBc3n4:bw63w0l) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L418)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

same answers. These derivations are (mostly) purely geometrical and do not perform the same operations as the `Shift()` operations. The chances of getting a match are real low.
</blockquote></details>

Fair point.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

edrumwri commented:
A few open comments and then <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 7 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 131 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knp5r2JfUqF6qrWNXj0:-Knp5r2JfUqF6qrWNXj1:bqrexvj) ([raw file](https://github.com/robotlocomotion/drake/blob/d964df95540b638fbd4ed8df80280e49516b47de/drake/multibody/multibody_tree/math/spatial_acceleration.h#L131)):*
> ```objc
>   /// <h3> Derivation </h3>
>   ///
>   /// <h4> Translational acceleration component </h4>
> ```

BTW Thanks for this description / derivation. It is indeed very helpful for my review- it helps me (and will likely help others) make explicit the connection between mechanics and the monogram notation.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 134 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knp6dMnqHePEOJWhA3J:-Knp6dMnqHePEOJWhA3K:b-is9b5y) ([raw file](https://github.com/robotlocomotion/drake/blob/d964df95540b638fbd4ed8df80280e49516b47de/drake/multibody/multibody_tree/math/spatial_acceleration.h#L134)):*
> ```objc
>   ///
>   /// The translational velocity `v_WBo` of point `Bo` in W can be obtained by
>   /// the shift operation as: <pre>
> ```

Please add that `Bo` is rigidly attached to frame P, which moves with angular velocity w_WP (measured from W) and translational velocity v_WPo (measured from the origin of W, Po). I know you take these definitions for granted now, but I don't expect the average reader to do so.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-KnpX7IFFX9SLXaKLuZq:b-ktndto)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Checkpoint: reviewed easy stuff and replied to some of Evan's comments.

---

Reviewed 2 of 7 files at r1, 1 of 7 files at r2, 2 of 7 files at r3.
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knd_NW4EP0PF-hoRinL:-KnpVlHRbdHyUbbujxcj:bg152z6) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L16)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I don't remember settling on any such wording for this. Can you justify it better than "that was the best we came up with six months ago"?
</blockquote></details>

BTW I think the "pairs of ordinary 3-vectors" part is worthwhile to distinguish these from Plücker vectors. The elementwise 0-2 and 3-5 part seems unnecessary here since that is a property of all our spatial vectors. But ... at least it's correct!

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 29 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne11WYR-GXM4l134SO:-KnpWcWznENaiUnf4B9S:b-ny3wmp) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L29)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

No, see the `ShiftTimeDerivative()` methods and unit tests first. Those are two different things. Also, take a look at `operator+()` for spatial acceleration.
</blockquote></details>

BTW operator+ is gone now -- it's ComposeWithMovingFrameAcceleration().

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1q2hhvOjDRW45MMB:-KnpWwJ_ZmHoYCoZV_he:b-n7d0e4) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L30)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Correct only for "measured in", not expressed in, yes?
</blockquote></details>

BTW the "frame fixed to a rigid body" language is accurate -- orthogonal to what purpose a frame is used.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1xRz786d13-90mys:-KnpXB9zjyX40SeWHmWl:bwtkenr) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L31)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I'm still left with a sense of ambiguity after reading this statement. To clarify, you mean "refers to a particular point affixed to a rigid body"? 
</blockquote></details>

BTW consider: "angular acceleration is identical for all frames fixed to the same rigid body".

---

*[drake/multibody/multibody_tree/math/spatial_vector.h, line 178 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbSH00u-Y0gQ22yRVE:-KnpYFHOCBuswp072k6G:b6zm7vn) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_vector.h#L178)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

No, I mean the braces operator.
</blockquote></details>

BTW the braces are part of the ["uniform initialization"](https://mbevin.wordpress.com/2012/11/16/uniform-initialization) feature set in C++11. In this case it is just invoking the default constructor as it would for `SpatialQuantity()`. For built-ins initializing with braces causes zero-initialization rather than default construction.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-Knp_4pt3EIzczj3jFEB:b-4q9pge)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 16 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knd_NW4EP0PF-hoRinL:-KnpUDBBomau5ImcZn8R:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L16)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I don't remember settling on any such wording for this. Can you justify it better than "that was the best we came up with six months ago"?
</blockquote></details>

Done.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1q2hhvOjDRW45MMB:-KnpXakPltqL_UCcBKYz:bev47i7) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L30)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Correct only for "measured in", not expressed in, yes?
</blockquote></details>

Could you elaborate on this one? I don't understand what you mean.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1xRz786d13-90mys:-KnpY1LMIefHpOULjVyZ:bpwo7dw) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L31)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

I'm still left with a sense of ambiguity after reading this statement. To clarify, you mean "refers to a particular point affixed to a rigid body"? 
</blockquote></details>

Actually, I just removed this sentence. I don't think rigid body mechanics should lead into frames' kinematics. It's ok as examples, but not as part of this class' doc.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 67 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbVVRcZUfuasWgg2ST:-KnpZQ74DTxHMeRgwwmY:b-co799l) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L67)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

This seems like the ideal place to talk about the layout of the underlying spatial acceleration, assuming that you actually need to do that. If you don't, I'd just say here that the first three components of the six dimensional vector refer to angular acceleration and the next three components refer to linear acceleration.
</blockquote></details>

I agree. Moved it in here. 

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 131 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knp5r2JfUqF6qrWNXj0:-KnpZY38ChSonH3v5N9J:b-bp97jk) ([raw file](https://github.com/robotlocomotion/drake/blob/d964df95540b638fbd4ed8df80280e49516b47de/drake/multibody/multibody_tree/math/spatial_acceleration.h#L131)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Thanks for this description / derivation. It is indeed very helpful for my review- it helps me (and will likely help others) make explicit the connection between mechanics and the monogram notation.
</blockquote></details>

BTW, thanks, it actually helped me also going through the careful exercise of writing it down.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 134 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knp6dMnqHePEOJWhA3J:-Knpa4QHeagD6Uqq16SL:bdxugf9) ([raw file](https://github.com/robotlocomotion/drake/blob/d964df95540b638fbd4ed8df80280e49516b47de/drake/multibody/multibody_tree/math/spatial_acceleration.h#L134)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

Please add that `Bo` is rigidly attached to frame P, which moves with angular velocity w_WP (measured from W) and translational velocity v_WPo (measured from the origin of W, Po). I know you take these definitions for granted now, but I don't expect the average reader to do so.
</blockquote></details>

I feel like I was using a little bit of a wrong notation. Changed to `v_WPb` and added explanatory comments.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 33 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbTgJ0uRuT8tE3mhTZ:-KnpaxRScDxBy_999CFi:b1y3kr4) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L33)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Would you mind giving it a better explanation? If I've not been reviewing similar code (even though I must have, I don't recall this, and you have to expect the average reader of this code will be seeing this for the first time), I have no idea what a "traits specialization" is or why it's appearing in this file.
</blockquote></details>

BTW, added explanatory text in the generic declaration just before this.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 349 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHgm6ImW895fUscwg:-KnpbUHqVfMPyLeGw87g:b4f8kae) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L349)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

BTW Ok, now I'm being pedantic, so would you please say instead "the angular velocity vector w_AP_E is orthogonal to the x/y plane"? 
</blockquote></details>

BTW, no problem

---

*[drake/multibody/multibody_tree/math/spatial_vector.h, line 178 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KnbSH00u-Y0gQ22yRVE:-Knp0I8WKHEDL6I2ujED:b-b2un0) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_vector.h#L178)):*
<details><summary><i>Previously, edrumwri (Evan Drumwright) wrote…</i></summary><blockquote>

No, I mean the braces operator.
</blockquote></details>

Notice this already made it into master now. So this files has no changes.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Review status: 4 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1q2hhvOjDRW45MMB:-KnpfcH4HYjzVD5NKaLT:b-p1j7o9) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L30)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

Could you elaborate on this one? I don't understand what you mean.
</blockquote></details>

Evan or me?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429)*
<!-- Sent from Reviewable.io -->

amcastro-tri commented:
Thanks for looking @sherm1! +@sherm1 for platform.

---

Review status: 4 of 6 files reviewed at latest revision, 7 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1q2hhvOjDRW45MMB:-KnphtBOdFbXwVNsSLH7:b-9h1ln3) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L30)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Evan or me?
</blockquote></details>

That question was for @edrumwri 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-KnphxvjS2wwhqy1vgRQ:b-up8hmq)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Platform review: <img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/>. The derivations were very helpful. I pushed some doc fixes directly, and added a TODO for @mitiguy to add a few independently-developed unit tests. Everything looks right but this is very tricky code and worth double checking since almost all future Drake results depend on it!

Because @amcastro-tri and I are headed off for vacations I want this to go in today, to give us one day to make sure all is well. I unblocked a few of your comments @edrumwri which I think have now been addressed. If not, please fix directly in a different PR or open an issue for Alejandro to deal with on return.

---

Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 30 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Kne1q2hhvOjDRW45MMB:-Knpvljf3r1sJ3ftmvui:b-261z4n) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/spatial_acceleration.h#L30)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

That question was for @edrumwri 
</blockquote></details>

BTW I think this is ok now -- unblocking Evan.

---

*[drake/multibody/multibody_tree/math/spatial_acceleration.h, line 134 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-Knp6dMnqHePEOJWhA3J:-Knpw6zBRv0sI6FFq2p-:bqt2ooi) ([raw file](https://github.com/robotlocomotion/drake/blob/d964df95540b638fbd4ed8df80280e49516b47de/drake/multibody/multibody_tree/math/spatial_acceleration.h#L134)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I feel like I was using a little bit of a wrong notation. Changed to `v_WPb` and added explanatory comments.
</blockquote></details>

BTW assuming this is satisfactory -- unblocking for expedience.

---

*[drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc, line 378 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-KneHvY65CJhbkguUjNa:-KnpwOU8U3l_4j-RYhd1:b-8opwm) ([raw file](https://github.com/robotlocomotion/drake/blob/e3ec47d8bcbe8cb5170a5a74f4f2ab044b792489/drake/multibody/multibody_tree/math/test/spatial_algebra_tests.cc#L378)):*
<details><summary><i>Previously, amcastro-tri (Alejandro Castro) wrote…</i></summary><blockquote>

I cannot think of a basic text for this. Maybe the classic Resnick Halliday on an introductory course on physics?
</blockquote></details>

I added a TODO here to have @mitiguy add some independent tests. Since he wrote the book, that should count as a published problem! Unblocking.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-KnpwdWpTCiS93Ww105g:b-iez4bs)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Ready to merge pending CI.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6429#-:-Knpy-UpiN_ief5gX5rG:bs4anw4)*
<!-- Sent from Reviewable.io -->
